scanpy/preprocessing/_scrublet/core.py

```diff
@@ -0,0 +1,466 @@
+    """
+
+    manifold_sim_: NDArray[np.float64] = field(init=False)
+    """shape (n_doublets × n_features)
+    The single-cell "manifold" coordinates (e.g., PCA coordinates)
+    for simulated doublets. Nearest neighbors are found using
+    the union of `manifold_obs_` (see above) and `manifold_sim_`.
+    """
+
+    doublet_parents_: NDArray[np.intp] = field(init=False)
+    """(shape: n_doublets × 2)
+    Indices of the observed transcriptomes used to generate the
+    simulated doublets.
+    """
+
+    doublet_neighbor_parents_: list[NDArray[np.intp]] = field(init=False)
+    """(length: n_cells)
+    A list of arrays of the indices of the doublet neighbors of
+    each observed transcriptome (the ith entry is an array of
+    the doublet neighbors of transcriptome i).
+    """
+
+    def __post_init__(
+        self,
+        counts_obs: sparse.csr_matrix | sparse.csc_matrix | NDArray[np.integer],
+        total_counts_obs: NDArray[np.integer] | None,
+        n_neighbors: int | None,
+        random_state: AnyRandom,
+    ) -> None:
+        self._counts_obs = sparse.csc_matrix(counts_obs)
+        self._total_counts_obs = (
+            self._counts_obs.sum(1).A.squeeze()
+            if total_counts_obs is None
+            else total_counts_obs
+        )
+        self._n_neighbors = (
+            int(round(0.5 * np.sqrt(self._counts_obs.shape[0])))
+            if n_neighbors is None
+            else n_neighbors
+        )
+        self._random_state = get_random_state(random_state)
+
+    def simulate_doublets(
+        self,
+        *,
+        sim_doublet_ratio: float | None = None,
+        synthetic_doublet_umi_subsampling: float = 1.0,
+    ) -> None:
+        """Simulate doublets by adding the counts of random observed transcriptome pairs.
+
+        Arguments
+        ---------
+        sim_doublet_ratio
+            Number of doublets to simulate relative to the number of observed
+            transcriptomes. If `None`, self.sim_doublet_ratio is used.
+
+        synthetic_doublet_umi_subsampling
+            Rate for sampling UMIs when creating synthetic doublets.
+            If 1.0, each doublet is created by simply adding the UMIs from two randomly
+            sampled observed transcriptomes.
+            For values less than 1, the UMI counts are added and then randomly sampled
+            at the specified rate.
+
+        Sets
+        ----
+        doublet_parents_
+        """
+
+        if sim_doublet_ratio is None:
+            sim_doublet_ratio = self.sim_doublet_ratio
+        else:
+            self.sim_doublet_ratio = sim_doublet_ratio
+
+        n_obs = self._counts_obs.shape[0]
+        n_sim = int(n_obs * sim_doublet_ratio)
+
+        pair_ix = sample_comb((n_obs, n_obs), n_sim, random_state=self._random_state)
+
+        E1 = cast(sparse.csc_matrix, self._counts_obs[pair_ix[:, 0], :])
+        E2 = cast(sparse.csc_matrix, self._counts_obs[pair_ix[:, 1], :])
```

Comment:
Again, not sure if the goal was copy as much and refactor later or not, but these are very unhelpful variable names.

Reply:
That being said, this one _is_ internal so won't contribute to the code-diff.  It might be worth handling internal changes here, but also don't mind if it's done later.

Reply:
> Again, not sure if the goal was copy as much and refactor later or not
yes, it is

> but these are very unhelpful variable names.
I agree so much. Worst part of working with mathematicians.